lib/pure/net.nim

```diff
@@ -191,6 +191,31 @@ proc isDisconnectionError*(flags: set[SocketFlag],
     SocketFlag.SafeDisconn in flags and
       lastError.int32 in {ECONNRESET, EPIPE, ENETRESET}
 
+proc checkCloseError*(ret: cint) =
+  ## Asserts that the return value of close() or closeSocket() syscall
+  ## does not indicate a programming error (such as invalid descriptor).
+  ## This must only be used when an error has already occurred and
+  ## you are performing a cleanup.
+  ## Otherwise, error handling must be performed as usual.
+  ##
+  ## This procedure must be called right after perfoming the syscall. Example:
+  ##
+  ## .. code-block:: nim
+  ##
+  ##  let ret = someSysCall()
+  ##  if ret != 0:
+  ##    let errcode = osLastError()
+  ##    checkCloseError sock.closeSocket()
+  ##    raise newException(OSError, osErrorMsg(errcode))
+
+  if ret != 0:
+    let errcode = osLastError()
+    when useWinVersion:
+      doAssert(errcode.int32 notin {WSANOTINITIALISED, WSAENOTSOCK,
+                                    WSAEINPROGRESS, WSAEINTR, WSAEWOULDBLOCK})
```

Comment:
We can't raise here, this procedure is designed to be used when another syscall has already failed and we are preparing to raise an exception, but need to cleanup first (see the actual usage and our previous conversation).

So, to recap, this procedure should replace `discard` when we need to close a socket within a cleanup section to prevent leaking. These codes indicate that the handle itself is incorrect/not configured properly, so we assert that the returned code is not one of these.